### PR TITLE
always-return-value

### PR DESCRIPTION
**Pull request by LukasKorba**
- The API has been refactored to carry state and always broadcast a value, even in case of failed request
- ignore refresh request when one is already in flight

This code review checklist is intended to serve as a starting point for the author and reviewer, although it may not be appropriate for all types of changes (e.g. fixing a spelling typo in documentation).  For more in-depth discussion of how we think about code review, please see [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md).

# Author

- [x] Self-review: Did you review your own code in GitHub's web interface? Code often looks different when reviewing the diff in a browser, making it easier to spot potential bugs.
- [ ] Automated tests: Did you add appropriate automated tests for any code changes?
- [ ] Code coverage: Did you check the code coverage report for the automated tests?  While we are not looking for perfect coverage, the tool can point out potential cases that have been missed.
- [ ] Documentation: Did you update Docs as appropiate? (E.g [README.md](../blob/main/README.md), etc.)
- [x] Run the app: Did you run the app and try the changes? 
- [ ] Did you provide Screenshots of what the App looks like before and after your changes as part of the description of this PR? (only applicable to UI Changes)
- [x] Rebase and squash: Did you pull in the latest changes from the main branch and squash your commits before assigning a reviewer? Having your code up to date and squashed will make it easier for others to review. Use best judgement when squashing commits, as some changes (such as refactoring) might be easier to review as a separate commit.

# Reviewer

- [ ] Checklist review: Did you go through the code with the [Code Review Guidelines](../blob/main/CODE_REVIEW_GUIDELINES.md) checklist?
- [ ] Ad hoc review: Did you perform an ad hoc review?  _In addition to a first pass using the code review guidelines, do a second pass using your best judgement and experience which may identify additional questions or comments. Research shows that code review is most effective when done in multiple passes, where reviewers look for different things through each pass._
- [ ] Automated tests: Did you review the automated tests?
- [ ] Manual tests: Did you review the manual tests?_You will find manual testing guidelines under our [manual testing section](../blob/mater/docs/testing/manual_testing)_
- [ ] How is Code Coverage affected by this PR? _We encourage you to compare coverage befor and after your changes and when possible, leave it in a better place. [Learn More...](../blob/master/docs/testing/local_coverage.md)_
- [ ] Documentation: Did you review Docs, [README.md](../blob/master/README.md), [LICENSE.md](../blob/master/LICENSE.md), and [Architecture.md](../blob/master/docs/Architecture.md) as appropriate?
- [ ] Run the app: Did you run the app and try the changes? While the CI server runs the app to look for build failures or crashes, humans running the app are more likely to notice unexpected log messages, UI inconsistencies, or bad output data.

### CODE REVIEW

File: Sources/ZcashLightClientKit/Tor/TorClient.swift [hunk 1]
```diff
@@ -10,6 +10,7 @@ import libzcashlc
 
 public class TorClient {
     private let runtime: OpaquePointer
+    public var cachedValue: FiatCurrencyResult?
```
Comment (str4d) [inline Sources/ZcashLightClientKit/Tor/TorClient.swift:13]: This is fine for now, but I'm not sure we want to cache here in the long term. Also `cachedValue` is a confusing name given that `TorClient` will eventually be used for more than just exchange rate fetching.

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift [hunk 2]
```diff
@@ -513,14 +516,34 @@ public class SDKSynchronizer: Synchronizer {
...
     public func refreshExchangeRateUSD() {
+        // ignore refresh request when one is already in flight
+        if let latestState = tor?.cachedValue?.state, latestState == .fetching {
+            return
+        }
+        
+        // broadcast cached value but update the state
+        if let cachedValue = tor?.cachedValue {
+            var fetchingState = cachedValue
+            fetchingState.state = .fetching
```
Comment (str4d) [inline Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:527]: Does this update the state stored in `tor.cachedValue`? This code looks like it is only updating a copy of the state; if that's the case then the "ignore refresh request" above would be doing nothing. But maybe this is just confusing Swift syntax.

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift [hunk 3]
```diff
@@ -27,6 +27,7 @@ public class SDKSynchronizer: Synchronizer {
     
     let metrics: SDKMetrics
     public let logger: Logger
+    var tor: TorClient?
```
Comment (str4d) [inline Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:30]: This should result in a performance improvement to rate refreshing (as Tor will be keeping its circuits up and ready in the background) while the synchronizer isn't torn down (which IIRC occurs at app backgrounding). We'll also want this for when we start routing certain lightwalletd requests over Tor.

**Review by str4d — APPROVED** [review]
utACK 1ffeff0c4b3fd6c7a60e77c83a00850422c7cffa with a question. It would be nice to change the confusing variable name.

File: Sources/ZcashLightClientKit/Tor/TorClient.swift (hunk 1 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Tor/TorClient.swift:13]: refactored to `cachedFiatCurrencyResult`

File: Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift (hunk 2 above)
Comment (LukasKorba) [reply Sources/ZcashLightClientKit/Synchronizer/SDKSynchronizer.swift:527]: Good catch, tor must be updated as well. I fixed that and also for the error state that is a few lines down bellow this code.

**Pull request merged**